### ISSUE TITLE
Response Code query is in desperate need of index

### ISSUE THREAD

**Issue by cabal95** [SparkDevNetwork/Rock#4180]
### TL; DR
Truncating the ResponseCode column to 100 characters and adding an index improves the "is the response code still available" query by a factor of 1,500x.

### The Problem
The specific query I am talking about is here: https://github.com/SparkDevNetwork/Rock/blob/1.10.2/Rock/Communication/Medium/Sms.cs#L440

It generates the following SQL, where `p__linq__0` is the response code being checked and `p__linq__1` is a date of 30 days ago:

```
(@p__linq__0 nvarchar(4000),@p__linq__1 datetime2(7))SELECT
    CASE WHEN ( EXISTS (SELECT
        1 AS [C1]
        FROM [dbo].[CommunicationRecipient] AS [Extent1]
        WHERE (([Extent1].[ResponseCode] = @p__linq__0) OR (([Extent1].[ResponseCode] IS NULL) AND (@p__linq__0 IS NULL))) AND ([Extent1].[CreatedDateTime] > @p__linq__1)
    )) THEN cast(1 as bit) ELSE cast(0 as bit) END AS [C1]
    FROM  ( SELECT 1 AS X ) AS [SingleRowTable1]
```

Here is the problem. Because ResponseCode has no index, all rows matching the date range must be loaded into memory and then have a string comparison (or null comparison) check performed on them.

On our server, we used this query to determine how many rows that is:

```
SELECT COUNT(*) FROM [CommunicationRecipient]
WHERE [CreatedDateTime] > DATEADD(DAY, -30, GETDATE())
```

This gave us 379,000 rows. Another church that is having much more significant slowdown due to this reported 560,000 rows.

In order to test the exact speed of the original query, I used SSMS and provided the LINQ parameters manually. Please note, that my date is more than 30 days ago. This test was performed on our beta server which did not have perfectly up to date data, so I used a date that would produce 30-days worth of results.

```
DECLARE @p__linq__0 AS nvarchar(4000) = '@100'
DECLARE @p__linq__1 AS datetime2(7) = '2/12/2020'

SELECT
    CASE WHEN ( EXISTS (SELECT
        1 AS [C1]
        FROM [dbo].[CommunicationRecipient] AS [Extent1]
        WHERE (([Extent1].[ResponseCode] = @p__linq__0) OR (([Extent1].[ResponseCode] IS NULL) AND (@p__linq__0 IS NULL))) AND ([Extent1].[CreatedDateTime] > @p__linq__1)
    )) THEN cast(1 as bit) ELSE cast(0 as bit) END AS [C1]
    FROM  ( SELECT 1 AS X ) AS [SingleRowTable1]
```

This resulted in a 1,421ms query. The other church I mentioned was seeing warnings about this query taking 30s or more.

### Potential Solutions
I tried various changes to the query, including having it check if the communication type was SMS, to see if that sped things up. There was only a very marginal difference in speed.

The next thing I tried was to create an index for this column. This took some doing as the column is currently an `nvarchar(max)` column which can't be indexed.  For this test I modified the column (on our beta box only) to be `nvarchar(100)` instead, then created an index on it.

This sped up the query by, well... infinity. The new query took 0ms to run. Below is a screenshot of the client statistics reported for the query. `Trial 1` is without the index, `Trial 2` is with the index.  I re-ran the original query a couple times to ensure that everything was cached that could be cached so this isn't a case of "first query took a long time, second query was obviously faster".

![image](https://user-images.githubusercontent.com/1119863/79021736-2c981600-7b31-11ea-8ce2-840326c41940.png)

I checked on our server, we didn't have a single ResponseCode column with a length of more than 6 characters. As far as I know, that should be the same for every church. The only place I can find this being used is for its intended purpose of `@######` codes. That said, this would be a breaking change.

If pursued, I would suggest a migration that first updates all CommunicationRecipient rows and truncates any non-null ResponseCode values to 100 characters. Then modifies the column to be an `nvarchar(100)` column and then finally applies the non-clustered index.

Because this could cause timeouts during migration, this would probably need to be done in a post-upgrade job in such a way that it can gracefully make this change without breaking the upgrade process.

### Versions
* **Rock Version:** 10.2
* **Client Culture Setting:** en-us

**Comment by cabal95** [comment]
I meant to include this as well. It shows how much space such an index would take in comparison to the other indexes already on this table.

![image](https://user-images.githubusercontent.com/1119863/79022295-76353080-7b32-11ea-94c3-780b58257ae0.png)
